core-v1/src/main/scala/lens/VariantLens.scala

```diff
@@ -0,0 +1,269 @@
+    * extension method on this lens. The `.select` method uses macros to make sure that the `name` param is
+    * set properly.
+    *
+    * @param name the name of the field
+    * @param getter a function to extract the value of this field from the object
+    * TODO: Use HList to make this more safe, instead of relying on the macro to use this properly.
+    */
+  def field[C](
+    name: String,
+    getter: B => C,
+  ): VariantLens[A, C] = {
+    copy(
+      path = path.atField(name),
+      get = this.get.andThen(getter),
+    )
+  }
+
+  /**
+    * Used for accessing the value for a given key in this object.
+    *
+    * This requires that there is an [[Indexed]] definition for how to access a value from the
+    * indexable object returned by the current lens.
+    *
+    * @param key either a value of the appropriate key type OR a [[shapeless.Nat]] for accessing a
+    *            specifically typed value from a tuple or [[HList]]
+    */
+  def at[K : ValidDataPathKey, V](
+    key: K,
+  )(implicit
+    CI: VariantIndexed[B, K, V],
+  ): VariantLens[A, V] = {
+    copy(
+      path = path.atKey(key),
+      get = get.andThen(b => CI.get(b)(key)),
+    )
+  }
+
+  /**
+    * Filters the given set of keys from this indexable object.
+    *
+    * @note Unfortunately this does not work for [[HList]] the same way that [[at]] does.
+    *
+    * TODO: It may be possible to make this work for tuples / HList with varargs instead of [[NonEmptySet]].
+    */
+  def filterKeys[K : ValidDataPathKey, V : Semigroup](
+    keys: NonEmptySet[K],
+  )(implicit
+    CI: VariantIndexed[B, K, V],
+  ): VariantLens[A, V] = {
+    import VariantIndexedSyntax._
+    copy(
+      path = path.filterKeys(keys),
+      get = get.andThen(_.filterKeys(keys)),
+    )
+  }
+
+  /**
+    * Upcasts the lens.
+    *
+    * @note this would not be necessary if [[VariantLens]] used appropriate variance, however, variance for lenses
+    *       introduces a lot of issues.
+    */
+  def as[V](implicit ev: B <:< V): VariantLens[A, V] =
+    this.copy(get = this.get.andThen(ev))
+
+  /**
+    * Converts the result into an [[HList]] if possible.
+    *
+    * Uses recursive implicit resolution at compile-time to derive the appropriate heterogenious return type.
+    *
+    * @see [[Generic]]
+    */
+  def asHList[C >: B](implicit gen: Generic[C]): VariantLens[A, gen.Repr] = copy(get = this.get.andThen(gen.to(_)))
+
+  // TODO: Are these runtime down casts needed? Why not just use .asInstanceOf (with a safe wrapper)
+
+//  def asIterable[V](implicit ev: B <:< IterableOnce[V]): VariantLens.AsIterableBuilder[A, IterableOnce, V] =
+//    new VariantLens.AsIterableBuilder(this.copy(get = this.get.andThen(ev)))
+//
+//  def asMap[K, V](implicit ev: B <:< IterableOnce[(K, V)]): VariantLens.AsMapBuilder[A, IterableOnce, K, V] =
+//    new VariantLens.AsMapBuilder(this.copy(get = this.get.andThen(ev)))
```

Comment:
We need some unit tests in this code to determine whether I can remove this TODO.